File: scripts/gps_enu.py
'''
MIT License
Copyright (c) 2019 Michail Kalaitzakis

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''

import numpy as np


class GPS_utils:
	'''
		Converts a gps signal (longitude, latitude, height) to a local cartesian ENU system
		
		Use setENUorigin(lat, lon, height) to set the local ENU coordinate system
		Use geo2enu(lat, lon, height) to get position in the local ENU system
	'''

	def __init__(self):
		# Geodetic System WGS 84 axes
		self.a = 6378137
		self.b = 6356752.314245
		self.a2 = self.a * self.a
		self.b2 = self.b * self.b
		self.e2 = 1 - (self.b2 / self.a2)

		# Local ENU Origin
		self.latZero = None
		self.lonZero = None
		self.hgtZero = None
		self.xZero = None
		self.yZero = None
		self.zZero = None
		self.R = np.asmatrix(np.eye(3))

	def setENUorigin(self, lat, lon, height):
		# Save origin lat, lon, height
		self.latZero = lat
		self.lonZero = lon
		self.hgtZero = height

		# Get origin ECEF X,Y,Z
		origin = self.geo2ecef(self.latZero, self.lonZero, self.hgtZero)
		self.xZero = origin.item(0)
		self.yZero = origin.item(1)
		self.zZero = origin.item(2)
		self.oZero = np.array([[self.xZero], [self.yZero], [self.zZero]])

		# Build rotation matrix
		phi = np.deg2rad(self.latZero)
		lmd = np.deg2rad(self.lonZero)

		cPhi = np.cos(phi)
		cLmd = np.cos(lmd)
		sPhi = np.sin(phi)
		sLmd = np.sin(lmd)

		self.R[0, 0] = -sLmd
		self.R[0, 1] = cLmd
		self.R[0, 2] = 0
		self.R[1, 0] = -sPhi*cLmd
		self.R[1, 1] = -sPhi*sLmd
		self.R[1, 2] = cPhi
		self.R[2, 0] = cPhi*cLmd
		self.R[2, 1] = cPhi*sLmd
		self.R[2, 2] = sPhi

	def geo2ecef(self, lat, lon, height):
		phi = np.deg2rad(lat)
		lmd = np.deg2rad(lon)

		cPhi = np.cos(phi)
		cLmd = np.cos(lmd)
		sPhi = np.sin(phi)
		sLmd = np.sin(lmd)

		N = self.a / (np.sqrt(1 - self.e2*sPhi*sPhi))

		x = (N + height)*cPhi*cLmd
		y = (N + height)*cPhi*sLmd
		z = ((self.b2 / self.a2)*N + height)*sPhi

		return np.array([[x], [y], [z]])

	def ecef2enu(self, x, y, z):
		ecef = np.array([[x], [y], [z]])

		return self.R * (ecef - self.oZero)

	def geo2enu(self, lat, lon, height):
		ecef = self.geo2ecef(lat, lon, height)

		return self.ecef2enu(ecef.item(0), ecef.item(1), ecef.item(2))

File: lib/bsc_common/gps_enu.cpp
#include "include/gps_enu.h"

namespace bsc_common
{
/* states
	// Geodetic System WGS 84 Axes
	double a = 6378137;
	double b = 6356752.314245;
	double a2 = a * a;
	double b2 = b * b;
	double e2 = 1 - (b2 / a2);

	// local ENU
	double latZero = null;
	double lonZero = null;
	double altZero = null;
	double xZero = null;
	double yZero = null;
	double zZero = null;
	Eigen::Matrix<double, 3, 3> R;
	*/

GPS_ENU::GPS_ENU()
{
}

void GPS_ENU::setENUOrigin(double lat, double lon, double alt)
{
	//Save origin lat, lon, height
	this->latZero = lat;
	this->lonZero = lon;
	this->altZero = alt;

	//Get origin ECEF X, Y, Z
	Eigen::Matrix<double, 3, 1> origin = this->geo2ecef(this->latZero, this->lonZero, this->altZero);
	this->xZero = origin[0];
	this->yZero = origin[1];
	this->zZero = origin[2];
	this->oZero << this->xZero, this->yZero, this->zZero;

	// Build rotation matrix
	double phi = this->latZero * M_PI / 180.0;
	double lmd = this->lonZero * M_PI / 180.0;

	double cPhi = std::cos(phi);
	double cLmd = std::cos(lmd);
	double sPhi = std::sin(phi);
	double sLmd = std::sin(lmd);

	this->R(0, 0) = -sLmd;
	this->R(0, 1) = cLmd;
	this->R(0, 2) = 0;
	this->R(1, 0) = -sPhi * cLmd;
	this->R(1, 1) = -sPhi * sLmd;
	this->R(1, 2) = cPhi;
	this->R(2, 0) = cPhi * cLmd;
	this->R(2, 1) = cPhi * sLmd;
	this->R(2, 2) = sPhi;
}
Eigen::Matrix<double, 3, 1> GPS_ENU::geo2ecef(double lat, double lon, double alt)
{
	double phi = this->latZero * M_PI / 180.0;
	double lmd = this->lonZero * M_PI / 180.0;

	double cPhi = std::cos(phi);
	double cLmd = std::cos(lmd);
	double sPhi = std::sin(phi);
	double sLmd = std::sin(lmd);

	double N = this->a / std::sqrt(1 - this->e2 * sPhi * sPhi);

	double x = (N + alt) * cPhi * cLmd;
	double y = (N + alt) * cPhi * sLmd;
	double z = ((this->b2 / this->a2) * N + alt) * sPhi;

	Eigen::Matrix<double, 3, 1> ret;
	ret << x, y, z;

	return ret;
}
Eigen::Matrix<double, 3, 1> GPS_ENU::ecef2enu(double x, double y, double z)
{
	Eigen::Matrix<double, 3, 1> ecef(x, y, z);

	return R * (ecef - oZero);
}
Eigen::Matrix<double, 3, 1> GPS_ENU::geo2enu(double lat, double lon, double alt)
{
	Eigen::Matrix<double, 3, 1> ecef = geo2ecef(lat, lon, alt);

	return ecef2enu(ecef(0, 0), ecef(1, 0), ecef(2, 0));
}
} // namespace bsc_common